# 🧬 Reconstruction of the Limit Cycle $Z_\\infty$

This notebook reconstructs the algebraic cycle $Z_\\infty = \\lim_k Z_k$  
from the symbolic iterates $\\omega_k = \\sum_i a_i^{(k)} \\alpha_i$.

We use the coefficients from the last $k$ iterations to estimate the limit:
\\[
Z_\\infty = \\sum_i a_i^{(\\infty)} Z_i
\\]

These $a_i^{(\\infty)}$ are obtained by averaging the last few symbolic iterates.

In [2]:
import numpy as np
import pandas as pd

# Simulated ω_k expressed in the basis {α₁, α₂, α₃}
omega_coeffs = np.array([
    [1.0, 2.0, 1.0],
    [1.0, 2.0, 0.5],
    [1.0, 2.0, 0.25],
    [1.0, 2.0, 0.125],
    [1.0, 2.0, 0.0625],
    [1.0, 2.0, 0.03125],
])

# Estimate a_i^∞ using the average of the last few iterates
Z_infty_coeffs = omega_coeffs[-3:, :].mean(axis=0)

df = pd.DataFrame({
    'i': [1, 2, 3],
    'a_i^∞ (estimated)': Z_infty_coeffs
})
df


i  a_i^∞ (estimated)
0  1           1.000000
1  2           2.000000
2  3           0.072917

## ✨ Output (LaTeX-ready)

Use the estimated coefficients to define the final cycle:

```latex
\\[
Z_\\infty \\approx 1.000 \\cdot Z_1 + 2.000 \\cdot Z_2 + 0.073 \\cdot Z_3
\\]
This expression is symbolic and explicitly reconstructs the algebraic cycle that realizes the limit of the symbolic iteration. 